libraries/node-core-library/src/JsonFile.ts

```diff
@@ -39,7 +39,92 @@ export type JsonObject = any;
 export type JsonNull = null;
 
 /**
- * Options for JsonFile.stringify()
+ * Specifies the variant of JSON syntax to be used.
+ *
+ * @public
+ */
+export enum JsonSyntax {
+  /**
+   * Specifies the exact RFC 8259 format as implemented by the `JSON.parse()` system API.
+   * This format was designed for machine generated inputs such as an HTTP payload.
+   * It is not a recommend choice for human-authored files, because it does not support
+   * code comments.
+   *
+   * @remarks
+   *
+   * A well-known quote from Douglas Crockford, the inventor of JSON:
+   *
+   * "I removed comments from JSON because I saw people were using them to hold parsing directives,
+   * a practice which would have destroyed interoperability.  I know that the lack of comments makes
+   * some people sad, but it shouldn't.  Suppose you are using JSON to keep configuration files,
+   * which you would like to annotate.  Go ahead and insert all the comments you like.
+   * Then pipe it through JSMin before handing it to your JSON parser."
+   *
+   * @see {@link https://datatracker.ietf.org/doc/html/rfc8259 | RFC 8259}
+   */
+  Strict = 'strict',
+
+  /**
+   * `JsonSyntax.JsonWithComments` is the recommended format for human-authored config files.
+   * It is a minimal extension to `JsonSyntax.Strict` adding support for code comments
+   * using `//` and `/*`.
+   *
+   * @remarks
+   *
+   * VS Code calls this format `jsonc`, but it should not be confused with unrelated file formats
+   * and libraries that also use the name "JSONC".
+   *
+   * To fix VS Code syntax highlighting, add this setting:
+   * `"files.associations": { "*.json": "jsonc" }`
+   *
+   * To fix GitHub syntax highlighting, add this to your `.gitattributes`:
+   * `*.json linguist-language=JSON-with-Comments`
+   */
+  JsonWithComments = 'jsonWithComments',
+
+  /**
+   * JSON5 is a project that proposes a JSON-like format supplemented with ECMAScript 5.1
+   * notations for objects, numbers, comments, and more.
+   *
+   * @remarks
+   * Files using this format should use the `.json5` file extension instead of `.json`.
+   *
+   * JSON5 has substantial differences from JSON: object keys may be unquoted, trailing commas
+   * are allowed, and strings may span multiple lines.  Whereas `JsonSyntax.JsonWithComments` can
+   * be cheaply converted to standard JSON by stripping comments, parsing JSON5 requires a
+   * nontrivial algorithm that may not be easily available in some contexts or programming languages.
+   *
+   * @see {@link https://json5.org/ | JSON5 project website}
+   */
+  Json5 = 'json5'
+}
+
+/**
+ * Options for {@link JsonFile.parseString}, {@link JsonFile.load}, and {@link JsonFile.loadAsync}.
+ *
+ * @public
+ */
+export interface IJsonFileParseOptions {
+  /**
+   * Specifies the variant of JSON syntax to be used.
+   *
+   * @defaultValue
+   * `JsonSyntax.Json5`
```

Comment:
@elliot-nelson @iclanton @D4N14L @dmichon-msft 

This is a longstanding oversight in the `JsonFile` API:

- `JsonFile.save()` always writes strict-JSON or JSON-with-comments (if we are updating a file that already contains comments)
- `JsonFile.load()` will silently accept full JSON5 syntax, for example `{ key: 0x1234, }`  (unquoted key, hex literal, trailing comma)

In practice, we don't actually use JSON5 syntax in our config files, because `JsonFile.save()` would reset it, and because every syntax highlighter would display it as an error.  In my opinion, accepting it was more of a bug than an undocumented feature. 😅

I feel strongly that if config files want to use JSON5, then they must use the `.json5` file extension.

(A similar case can be made that JSON files with comments should use a file extension such as `.jsonc` or `.cjson`, however that argument is weaker because (1) nobody seems to agree about what this file extension should be and (2) for web projects, most `.json` files are human-authored and it's ***plainly misguided*** to forbid comments in human-authored files.)

### Proposal

In a separate PR, I propose to change the default for `IJsonFileParseOptions.jsonSyntax` to be `JsonSyntax.JsonWithComments`.  It would be a MAJOR version bump for `node-core-library`, however my suspicion is that it's unlikely to break any real-world projects using this library.

Thoughts?

Reply:
The following thoughts are likely out of scope of this bug, but discuss consistency issues amongst files with the `.json` extension in a given RushJS repo.

The disparity between the `rush.json` file being written in `cjson` syntax and every other `package.json` necessarily being written in strict `json` mode (largely due to the package managers' (at least `pnpm`'s) inability to parse plain `cjson`/`json5`) is a bit confusing in that it sets inconsistent format expectations for files with the same extension across different `.json` files in a given repo.

Anecdotally, I have observed our devs at Adobe add `cjson` syntax to `rush.json`. I imagine that it's not far-fetched to say that other projects may have made `json5` modifications to their respective `rush.json` since its parsing has been open `json5` by default. At this stage restricting `rush.json` to `cjson` would fail parsing in certain repos that have taken liberties due to suddenly stricter parsing rules of what may have effectively become a JSON5-formatted file. 

One potential avenue is to rename the `rush.json` file currently written in `cjson` format in the RushStack repo to `.json5` extension, explicitly opening the door to JSON5 format. I imagine there would be a deprecation warning when building a repo using the the older `rush.json` file, and a loading precedence or error necessary in case both a `rush.json` and `rush.json5` are found in a project.

Caveat: my suggestion would have been to stick with `cjson` format as opposed to `json5` for `rush.json` if a respective disparate file extension could be agreed upon for that file. In that case, an error for any repos with `json5` modifications to their `rush.json` would be shown.

Just my 2c.

Reply:
> Anecdotally, I have observed our devs at Adobe add single line comment syntax to `rush.json`.

This is valid `cjson` and heavily used in Rush's template problems.

> I imagine that it's not far-fetched to say that other projects may have made `json5` modifications to their respective `rush.json` since its parsing has been open `json5` by default. At this stage restricting `rush.json` to `cjson` would fail parsing in certain repos that have taken liberties due to suddenly stricter parsing rules of what may have effectively become a JSON5-formatted file.

It's not far-fetched, but my hypothesis is that it is very rare.  You can test it easily by modifying your branch to change this line to say `cjson`, then running `rush install` [using your dev build](https://rushjs.io/pages/contributing/#testing-rush-builds):

https://github.com/microsoft/rushstack/blob/8a497e11561d03b44a73675c50318f37c4e0db7c/libraries/node-core-library/src/JsonFile.ts#L590-L595

I tried it with several if our repos and there were no problems.  It makes sense, because nothing about the syntax highlighting or existing file contents would suggest that JSON5 syntaxes are allowable.  If there are JSON5 syntaxes in a few places, fixing them should be pretty easy.

Also keep in mind that Rush always parses config files using the engine version specified by `rushVersion` from **rush.json**.  Thus, upgrading to the stricter Rush would be an explicit action, and thereafter there won't be any concern about users invoking the old version of Rush.